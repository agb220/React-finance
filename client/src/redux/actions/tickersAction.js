import { io } from "socket.io-client";

const socket = io("http://localhost:4000");

socket.emit("start");

export const socketTickers = () => (dispatch, getState) => {
  socket.on("ticker", function (response) {
    //console.log("start");
    const data = Array.isArray(response) ? response : [response];

    dispatch(setTickers(data));

    const state = getState();
    //console.log("state", state);
    const history = data.reduce((accumulator, currentValue) => {
      // console.log("accumulator", accumulator, "currentValue", currentValue);

      if (accumulator[currentValue.ticker]) {
        accumulator[currentValue.ticker] = accumulator[
          currentValue.ticker
        ].concat([currentValue.price]);
      } else {
        accumulator[currentValue.ticker] = [currentValue.price];
      }
      return accumulator;
    }, state.tickersReducer.history);

    // console.log("data", data);

    dispatch(setTickersHistory(history));
  });
};

export const setTickers = (tickers) => ({
  type: "SET_TICKERS",
  payload: tickers,
});

export const setTickersHistory = (tickers) => ({
  type: "SET_TICKERS_HISTORY",
  payload: tickers,
});

export const setLoaded = (payload) => ({
  type: "SET_LOADED",
  payload,
});
