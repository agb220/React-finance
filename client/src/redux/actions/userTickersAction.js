export const setUserTickers = (tickers) => ({
  type: "SET_USER_TICKERS",
  payload: tickers,
});
