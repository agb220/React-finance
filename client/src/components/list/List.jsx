import React from "react";

import { useSelector, useDispatch } from "react-redux";

import styled from "styled-components";

import { setUserTickers } from "../../redux/actions/userTickersAction";

import Loader from "../loader/Loader";
import Item from "./Item";

const Wrapper = styled.section`
  width: 100%;
  padding: 1rem 0;
  display: flex;
  flex-direction: column;
`;

const Title = styled.h1`
  font-size: var(--fs-lg);
`;

function List() {
  const dispatch = useDispatch();
  const tickers = useSelector(({ tickersReducer }) => tickersReducer.tickers);
  const isLoaded = useSelector(({ tickersReducer }) => tickersReducer.isLoaded);

  const handleAddUserTickers = (tickers) => {
    dispatch(setUserTickers(tickers));
  };

  return (
    <>
      <Wrapper>
        <Title>List tickers</Title>
        {isLoaded
          ? tickers &&
            tickers.map((ticker, index) => (
              <Item
                ticker={ticker}
                key={ticker.ticker + index}
                onClickAddTickers={handleAddUserTickers}
              />
            ))
          : Array(10)
              .fill(0)
              .map((_, index) => <Loader key={index} />)}
      </Wrapper>
    </>
  );
}

export default List;
