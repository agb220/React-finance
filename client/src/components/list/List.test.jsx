import { render, screen, fireEvent } from "@testing-library/react";
import renderer from "react-test-renderer";
import "@testing-library/jest-dom";
import "jest-styled-components";

import styled from "styled-components";
import * as reduxHooks from "react-redux";
import * as actions from "../../redux/actions/tickersAction";
import * as userActions from "../../redux/actions/userTickersAction";

import List from "./List";

jest.mock("react-redux");

const mockedSelector = jest.spyOn(reduxHooks, "useSelector");
const mockedDispatch = jest.spyOn(reduxHooks, "useDispatch");

const ticker = [
  {
    ticker: "AAPL",
    exchange: "NASDAQ",
    price: "190.67",
    change: "155.50",
    change_percent: "0.62",
    dividend: "0.91",
    yield: "1.86",
    last_trade_time: "2023-03-07T04:27:51.000Z",
  },
];

describe("List component", () => {
  test("Should create List component with empty ticker", () => {
    mockedSelector.mockReturnValue([]);
    const component = render(<List />);
    expect(component.firstChild).toMatchSnapshot();
  });
  test("Should create List component with ticker", () => {
    mockedSelector.mockReturnValue(ticker);
    const component = render(<List />);
    expect(component.firstChild).toMatchSnapshot();
  });
  test("Should create List component", () => {
    mockedDispatch.mockReturnValue(jest.fn());
    const component = render(<List />);
    expect(component.firstChild).toMatchSnapshot();
  });

  // test("Should dispatch actions", () => {
  //   const dispatch = jest.fn();
  //   mockedDispatch.mockReturnValue(dispatch);
  //   const mockedSetUserTickers = jest.spyOn(actions, "setTickers");

  //   render(<List ticker={ticker} />);

  //   // fireEvent.click(screen.getRole("add"));

  //   expect(dispatch).toHaveBeenCalledTimes(1);
  //   expect(mockedSetUserTickers).toHaveBeenCalled();
  // });
});
