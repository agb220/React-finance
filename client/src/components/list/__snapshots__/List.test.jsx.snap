// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List component Should create List component 1`] = `undefined`;

exports[`List component Should create List component with empty ticker 1`] = `undefined`;

exports[`List component Should create List component with ticker 1`] = `undefined`;
